gpcontrib/arenadata_toolkit/src/track_files.c

```diff
@@ -0,0 +1,1228 @@
+		}
+		/* initial snapshot shouldn't return drops */
+		if (tf_get_global_state.bloom && !tf_get_global_state.bloom->is_set_all)
+		{
+			tf_get_global_state.drops = drops_track_move(MyDatabaseId);
+			tf_get_global_state.next_drop = list_head(tf_get_global_state.drops);
+		}
+
+		/*
+		 * Let's retrieve tracking information only once for the transaction.
+		 */
+		if (tf_get_global_state.schema_oids == NIL)
+			get_filters_from_guc();
+
+		if (tf_get_global_state.relstorages == NIL ||
+			tf_get_global_state.relkinds == NIL ||
+			tf_get_global_state.schema_oids == NIL)
+			elog(ERROR, "cannot get tracking configuration (schemas, relkinds, reltorage) for database %u", MyDatabaseId);
+
+		MemoryContextSwitchTo(oldcontext);
+
+		oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
+
+		funcctx->tuple_desc = CreateTemplateTupleDesc(9, false);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber)1, "relid", OIDOID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber)2, "name", NAMEOID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber)3, "relfilenode", OIDOID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber)4, "size", INT8OID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber)5, "state", CHAROID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber)6, "gp_segment_id", INT4OID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber)7, "relnamespace", OIDOID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber)8, "relkind", CHAROID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber)9, "relstorage", CHAROID, -1, 0);
+		funcctx->tuple_desc = BlessTupleDesc(funcctx->tuple_desc);
+
+		state = (tf_main_func_state_t *) palloc0(sizeof(tf_main_func_state_t));
+		funcctx->user_fctx = (void *)state;
+
+		if (tf_get_global_state.bloom)
+		{
+			state->pg_class_rel = heap_open(RelationRelationId, AccessShareLock);
+			state->scan = systable_beginscan(state->pg_class_rel, InvalidOid, false, NULL, 0, NULL);
+		}
+
+		MemoryContextSwitchTo(oldcontext);
+	}
+
+	funcctx = SRF_PERCALL_SETUP();
+	state = funcctx->user_fctx;
+
+	while (true)
+	{
+		Oid			filenode;
+		Oid			relnamespace;
+		char		relkind;
+		char		relstorage;
+		HeapTuple	pg_class_tuple;
+		uint64_t	hash;
+
+		if (!state->scan)
+			break;
+
+		pg_class_tuple = systable_getnext(state->scan);
+
+		if (!HeapTupleIsValid(pg_class_tuple))
+		{
+			systable_endscan(state->scan);
+			heap_close(state->pg_class_rel, AccessShareLock);
+			state->scan = NULL;
+			state->pg_class_rel = NULL;
+			break;
+		}
+
+		datums[6] = heap_getattr(pg_class_tuple, Anum_pg_class_relnamespace, RelationGetDescr(state->pg_class_rel), &nulls[6]);
+		relnamespace = DatumGetObjectId(datums[6]);
+
+		if (!schema_is_tracked(relnamespace))
+			continue;
+
+		datums[7] = heap_getattr(pg_class_tuple, Anum_pg_class_relkind, RelationGetDescr(state->pg_class_rel), &nulls[7]);
+		relkind = CharGetDatum(datums[7]);
```

Comment:
it should be `DatumGetChar`
```suggestion
		relkind = DatumGetChar(datums[7]);
```
The same below.

Reply:
Have changed